File: gobackN.py
# Define the total number of frames and window size
total_frames = int(input("Enter the Total number of frames: "))
window_size = int(input("Enter the Window Size: "))

sent_frames = []
transmitted_frames = 0
current_frame = 1

while current_frame <= total_frames:
    acknowledged_count = 0

    # Send frames within the current window
    for frame in range(current_frame, current_frame + window_size):
        print(f"Sending Frame {frame}...")
        sent_frames.append(frame)
        transmitted_frames += 1

    # Receive acknowledgments or handle timeouts
    for frame in range(current_frame, current_frame + window_size):
        received = int(input(f"Did you receive frame {frame}?\n1 for yes, 0 for no: "))

        if received == 1:
            print(f"Acknowledgment for Frame {frame}...")
            acknowledged_count += 1
        else:
            print(f"Timeout!! Frame Number: {frame} Not Received")
            print("Retransmitting Window...")
            break

    print()
    current_frame += acknowledged_count

    # Remove acknowledged frames from the sent_frames list
    sent_frames = sent_frames[acknowledged_count:]

print(f"Total number of frames sent and resent: {transmitted_frames}")

File: GoBackN.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class GoBackN {
    public static void main(String[] args) {
        //1, 2, 3, 4, 5
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the Total number of frames: ");
        int tf = scanner.nextInt();
        System.out.print("Enter the Window Size: ");
        int N = scanner.nextInt();

        List<Integer> sentFrames = new ArrayList<>();
        int i = 1;
        int tt = 0; //tt (total transmitted frames)

        while (i <= tf) { //i =3
            int z = 0; //z is used to count the number of frames that were successfully acknowledged within the current window.
            for (int k = i; k < i + N && k <= tf; k++) {
                System.out.println("Sending Frame " + k + "...");
                sentFrames.add(k);
                tt++;
            }

            for (int k = i; k < i + N && k <= tf; k++) {
                int f;
                System.out.println(k);
                System.out.println("Did you receive frame " + k + "?");
                System.out.println("1 for yes, 0 for no");
                f = scanner.nextInt();

                if (f == 1) {
                    System.out.println("Acknowledgment for Frame " + k + "...");
                    z++;
                } else {
                    System.out.println("Timeout!! Frame Number : " + k + " Not Received");
                    System.out.println("Retransmitting Window...");
                    break;
                }
            }

            System.out.println();
            i = i + z;

            // Remove acknowledged frames from the sentFrames list
            for (int j = 0; j < z; j++) {
                sentFrames.remove(0);
            }
        }

        System.out.println("Total number of frames which were sent and resent are: " + tt);
    }
}
